src/Orleans.Runtime/Core/Dispatcher.cs

```diff
@@ -343,37 +343,33 @@ public bool CanInterleave(ActivationData targetActivation, Message incoming)
         /// during duration of request processing. If target of outgoing request is found in that collection - 
         /// such request will be marked as interleaving in order to prevent deadlocks.
         /// </summary>
-        private void MarkSameCallChainMessageAsInterleaving(ActivationData sendingActivation, Message outgoing)
+        private bool IsCallChainReentrancyAllowed(ActivationData targetActivation, Message incoming)
         {
-            if (!schedulingOptions.AllowCallChainReentrancy)
+            if (!schedulingOptions.AllowCallChainReentrancy
+                 || incoming.Direction == Message.Directions.OneWay)
             {
-                return;
+                return false;
             }
 
-            if (outgoing.Direction == Message.Directions.OneWay)
+            if (incoming.CallChainId == null || targetActivation.Running == null)
             {
-                return;
+                return false;
             }
 
-            if (sendingActivation?.RunningRequestsSenders.Contains(outgoing.TargetActivation) == true)
-            {
-                outgoing.IsAlwaysInterleave = true;
-            }
+            var isCallFromOriginatorActivation = incoming.SendingActivation.Equals(targetActivation.Running.SendingActivation);
```

Comment:
I don't quite understand this condition, could you explain the reasoning with a comment?

Reply:
Is the point here to avoid immediate reentrancy, eg in the case of `this.AsReference<TSelf>().SomeMethod().Ignore()`?

That behavior may be worthy of its own option, but I suppose 'allow call chain reentrancy' doesn't preclude some cases where reentrancy within a chain is disallowed.

Reply:
Sorry for late reply; the point here is to avoid reentrancy for multiple requests which are coming from single origin (fan out to single target ).